include/kernels/staggered_quark_smearing.cuh

```diff
@@ -0,0 +1,229 @@
+      is_t0_kernel(is_t0_kernel),
+      t0_offset(is_t0_kernel ? in.VolumeCB() / in.X(3) : 0)
+    {
+      if (in.V() == out.V()) errorQuda("Aliasing pointers");
+      checkOrder(out, in);        // check all orders match
+      checkPrecision(out, in, U); // check all precisions match
+      checkLocation(out, in, U);  // check all locations match
+      if (!in.isNative() || !U.isNative())
+        errorQuda("Unsupported field order colorspinor(in)=%d gauge=%d combination\n", in.FieldOrder(), U.FieldOrder());
+      if (dir < 3 || dir > 4) errorQuda("Unsupported laplace direction %d (must be 3 or 4)", dir);
+
+      for( int i=0; i<4; i++ )
+      {
+        t0_face_offset[i] = is_t0_kernel ? (int)(this->dc.face_XYZ[i])/2 : 0;
+        face_size[i] = 3 * this->dc.ghostFaceCB[i]; // 3=Nface
+        t0_face_size[i] = ( is_t0_kernel && i<3 ) ? face_size[i] / in.X(3) : face_size[i];
+      }
+
+      // partial replication of dslash::setFusedParam()
+      int prev = -1;
+      for( int i=0; i<4; i++ )
+      {
+        threadDimMapLower_t0[i] = 0;
+        threadDimMapUpper_t0[i] = 0;
+        if( !(this->commDim[i]) ) continue;
+        threadDimMapLower_t0[i] = ( prev>=0 ? threadDimMapUpper_t0[prev] : 0 );
+        threadDimMapUpper_t0[i] = threadDimMapLower_t0[i] + 2 * t0_face_size[i];
+        prev = i;
+      }
+    }
+  };
+
+  /**
+     Applies the off-diagonal part of the covariant derivative operator
+
+     @param[out] out The out result field
+     @param[in,out] arg Parameter struct
+     @param[in] U The gauge field
+     @param[in] coord Site coordinate struct
+     @param[in] parity The site parity
+     @param[in] idx Thread index (equal to face index for exterior kernels)
+     @param[in] thread_dim Which dimension this thread corresponds to (fused exterior only)
+
+  */
+  template <int nParity, KernelType kernel_type, int dir, typename Coord, typename Arg, typename Vector>
+  __device__ __host__ inline void applyStaggeredQSmear(Vector &out, Arg &arg, Coord &coord, int parity,
+                                               int, int thread_dim, bool &active)
+  {
+    typedef typename mapper<typename Arg::Float>::type real;
+    typedef Matrix<complex<real>, Arg::nColor> Link;
+    const int their_spinor_parity = (arg.nParity == 2) ? parity : 0;
+
+#pragma unroll
+    for (int d = 0; d < Arg::nDim; d++) { // loop over dimension
+      if (d != dir) {
+        {
+          // Forward gather - compute fwd offset for vector fetch
+          const bool ghost = (coord[d] + 2 >= arg.dim[d]) && isActive<kernel_type>(active, thread_dim, d, coord, arg);//1=>2
+	  
+          if (doHalo<kernel_type>(d) && ghost) {//?
+
+            const int ghost_idx = ghostFaceIndexStaggered<1>(coord, arg.dim, d, 2);//check nFace=2, requires improved staggered fields
+            const Link U = arg.U(d, coord.x_cb, parity);
+            const Vector in = arg.in.Ghost(d, 1, ghost_idx, their_spinor_parity);//?
+
+            out += U * in;
```

Comment:
I think this line, and equivalently on lines 127, 141 and 150 should be replaced with a call to `mv_add`.  This will ensure the operations are performed using the optimal number of `fma` instructions.  See for example [here](https://github.com/lattice/quda/blob/develop/include/kernels/dslash_staggered.cuh#L105).

Reply:
done (no difference in performance, though)